packages/flutter/lib/src/material/text_field.dart

```diff
@@ -168,6 +186,38 @@ class TextField extends StatefulWidget {
   /// null, the value must be greater than zero.
   final int maxLines;
 
+  /// The maximum number of characters (Unicode runes) to allow in the text field.
+  ///
+  /// If set, a character counter will be displayed below the
+  /// field, showing how many characters have been entered and how many are
+  /// allowed. After [maxLength] characters have been input, additional input
+  /// is ignored, unless [maxLengthEnforced] is set to false. The TextField
+  /// enforces the length with a [LengthLimitingTextInputFormatter], which is
+  /// evaluated after the supplied [inputFormatters], if any.
+  ///
+  /// This value must be either null or greater than zero. If set to null
+  /// (the default), there is no limit to the number of characters allowed.
+  ///
+  /// Whitespace characters (e.g. newline, space, tab) are included in the
+  /// character count.
+  ///
+  /// If [maxLengthEnforced] is set to false, then more than [maxLength]
+  /// characters may be entered, but the error counter and divider will
+  /// switch to the [decoration.errorStyle] when the limit is exceeded.
+  ///
+  /// See also:
+  ///  * [LengthLimitingTextInputFormatter] for more information on how it
+  ///    counts characters.
+  final int maxLength;
```

Comment:
I think we should probably duplicate the information here, and give specific examples of what characters are (e.g. showing the difference between composed and decomposed characters, emoji sequences, and so on).

Reply:
Specifically I meant the disclaimer about what a "character" is (e.g. that this will break with non-trivial emoji, e.g. those with skin tone selectors, and that therefore one should be generous with the length).